data_engineering/derive/derive.py

```diff
@@ -0,0 +1,71 @@
+import pandas as pd
+
+from data_engineering.derive.fourth_layer_derivation import derive_fourth_layer
+from data_engineering.derive.second_layer_derivation import derive_second_layer
+from data_engineering.derive.third_layer_derivation import derive_third_layer
+
+
+def derive_total_assets_from_log_total_assets(df: pd.DataFrame) -> pd.DataFrame:
+    return df.assign(**{"TOTAL_ASSETS": 10 ** df["logarithm of total assets"],})
+
+
+def derive_short_term_securities_from_cash(df: pd.DataFrame) -> pd.DataFrame:
+    return df.assign(
+        **{
+            "SHORT_TERM_SECURITIES": (
+                df[
+                    "[(cash + short-term securities + receivables - short-term liabilities) / (operating expenses - depreciation)] * 365"
+                ]
+                / 365
+            )
+            * (df["OPERATING_EXPENSES"] - df["DEPRECIATION"])
+            - df["CASH"]
+            - df["RECEIVABLES"]
+            + df["SHORT_TERM_LIABILITIES"]
+        }
+    )
+
+
+def calculate_number_of_null_values_per_row(df):
+    return df.assign(**{"NULL_VALUE_COUNT": df.isna().sum(axis=1)})
+
+
+def remove_companies_with_null_values(df):
+    # todo: maybe we should explicitly remove the row using that company_id? or use the assign above to help filter
+    return df.loc[(~df["TOTAL_ASSETS"].isna())]
+
+
+def remove_duplicates(df):
+    # todo: need to try different combination of cols to find other possible duplicates
+    cols = list(df.columns)
+    cols.remove("COMPANY_ID")
+    return df.drop_duplicates(cols)
+
```

Comment:
Are these two function not used?

Reply:
im not sure we should remove duplicates/null values before getting the raw values or after getting the raw values. the functions will be used but I want to discuss with others on when to use it. what do you think?

Reply:
yeah I was thinking about the same thing. This would depends on if the derivative operation will introduce more null... worth discussion for sure!

Reply:
I extracted these functions to a new file called clean.py within utils. Will decide when to use it after EDA